### ISSUE TITLE
if a build step in tekton pipeline fails, it should not go on to the next build step

### ISSUE THREAD

**Issue by tam512** [kabanero-io/kabanero-pipelines#86]
Using Kabanero 0.2.0 rc1, running a java-microprofile-build-deploy-pipeline in tekton dashboard. There was a problem in the pom.xml which caused the build-bud step fails (red mark), but it continues to go on to the next step build-push and show as green completed with `No log available`
1) If a build step fails, it should not go on to the next build step.
2) If it has to go on to the next step, then it should be failed also

![image](https://user-images.githubusercontent.com/22480038/66674949-e584a300-ec29-11e9-8d86-2da7b1f65869.png)

![image](https://user-images.githubusercontent.com/22480038/66674966-f0d7ce80-ec29-11e9-98ad-da9bb4648e65.png)

**Comment by aadeshpa** [comment]
@kalathur: I think we should we using `runAfter: [step-name]` like we have used for the tasks in our `build-deploy-pipeline.yaml` file.

**Comment by kvijai82** [comment]
My first reaction without reproducing this is that this is a tekton dashboard issue.  I don't think we actually run the build-push step.  It would have blown up with lots of errors if it did.  So I don't think we have a major issue here.  I have only seen 2 states for steps in the dashboard.  The red cross or the green check.  Not sure if there is a 3rd one representing didn't execute.  We will investigate it a bit more today to see how it looks like from the command line view today and decide on the next steps.

**Comment by aadeshpa** [comment]
We were able to re-produce this issue as below

1. I created webhook for java-microprofile project in my `github.ibm.com` from the tekton dashboard.
2. I made some code changes with `compile time errors` , so the `build-bud` step fails.

Results:

- As the pipelinerun failed, the tekton dashboard shows exact screenshot as per this issue.

- However when we investigated from the oc logs command we found below results

1. `oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-bud`

Logs Snippet for `build-bud` container in the pod:-

```
Downloaded from central: https://repo.maven.apache.org/maven2/com/google/collections/google-collections/1.0/google-collections-1.0.jar (640 kB at 1.2 MB/s)
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to /project/user-app/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[8,5] <identifier> expected
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[9,2] reached end of file while parsing
[INFO] 2 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57.022 s
[INFO] Finished at: 2019-10-15T15:27:33Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project starter-app: Compilation failure: Compilation failure: 
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[8,5] <identifier> expected
[ERROR] /project/user-app/src/main/java/dev/appsody/starter/StarterApplication.java:[9,2] reached end of file while parsing
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
subprocess exited with status 1
subprocess exited with status 1
error building at STEP "RUN mvn install -DskipTests": exit status 1
```
2. `oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-push`

Logs never returned anything 👎 
```
[root@crofts1 common]# oc logs -f my-java-mp-webhook-1571152644-build-task-wndcp-pod-2f0e7d -c step-build-push
[root@crofts1 common]# 
```

Conclusion:
This means that `build-bud` container step failed, and the `build-push` never started. However the tekton-dashboard shows the `build-push` step got completed. So we think it is a Tekton Dashboard issue.

**Comment by aadeshpa** [comment]
This issue is marked as blocked by tekton and we have opened the issue in Tekton Dashboard repository:-
https://github.com/tektoncd/dashboard/issues/644

**Comment by stephenkinder** [comment]
Unlikely that Tekton would address in time to make release 0.2.0 and I removed the tag.  We should pursue the issue under the tekton dashboard repo.

**Comment by bhatsuresh** [comment]
This would be a case generator in my mind from the field and one where
folks could go "do you guys even test this"... :-)

Lets pursue the fix or a mitigation please...the build push should not be
reflecting green on the dashboard or it should just stop on exiting the
failure from the previous step.

Thanks
---------------------------------------------------------------------------------------------------

Suresh Bhat
Program Director, WAS-Dev.Cloud Subtribe, Performance and SVT
RTP, Durham, North Carolina
ph: 919.543.6068
slbhat@us.ibm.com
------------------------------------------------------------------------

**Comment by gecock** [comment]
@aadeshpa @stephenkinder it looks like Tekton fixed their issue.  Will this be making it into the next Kabanero release?

**Comment by kvijai82** [comment]
@gecock from what In understand, this fix in only in tekton 0.8, which we are not picking up in kabanero 0.3.0.  We can look into pulling this in for 0.4.0.

**Comment by kvijai82** [comment]
I don't think we are going to be able to contain pulling in tekton 0.8 in 0.4.0 with the time we have lost working on 0.3.0 issues in 0.4.0.  Might have to wait another iteration for this.

**Comment by marikaj123** [comment]
@mtamboli - Are you good if I update the Kabanero release to v0.5.0 based on the comments provided by Vijai above?

**Comment by marikaj123** [comment]
@kvijai82 - Is Kabanero 0.5.0 pulling tekton 0.8?

**Comment by kvijai82** [comment]
As far as I know the plan is to pull in tekton 0.8 into Kabanero 0.5.0.  We
were able to successfully test triggers.

**Comment by aadeshpa** [comment]
I verified with collections 0.4.0 release with below setup
1. Created webhook for Java microprofile collection project in github.
2. Edited java file with syntax error, and commited the changes, which triggerred build-push pipeline
3. The pipelinerun failed in build step as expected and in the dashboard it showed that nothing ran after the build step.

This shows if a step fails the pipelinerun in dashboard does not go to next step.
<img width="1419" alt="Screen Shot 2020-01-10 at 3 54 53 PM" src="https://user-images.githubusercontent.com/53223993/72185660-5706f580-33c1-11ea-9daa-8e6945116577.png">

**Comment by marikaj123** [comment]
@tam512 - According to Ashish the issue is not reproducible, can you please test it with the latest Tekton versions?

**Comment by tam512** [comment]
I'm testing with Kabanero 0.5.0-rc2 which includes Tekton version v0.8.0 and I do not see build-bud step anymore
<img width="913" alt="Screen Shot 2020-01-15 at 4 42 33 PM" src="https://user-images.githubusercontent.com/22480038/72477923-55b93c80-37b6-11ea-8d00-caaebcb490aa.png">

**Comment by ncskier** [comment]
@tam512 I just verified that I also see the bug when running Tekton Dashboard v0.3.0 and Tekton Pipelines v0.8.0. 
![Screen Shot 2020-01-16 at 11 53 37 AM](https://user-images.githubusercontent.com/16961496/72545318-dc255b00-3856-11ea-9f8f-a601ade79337.png)

This bug was fixed in the Tekton Dashboard v0.2.1 with PR https://github.com/tektoncd/dashboard/pull/653. However, the fix PR was reverted for Dashboard v0.3.0 when Tekton Pipelines v0.9.0 fixed an underlying issue causing the bug https://github.com/tektoncd/dashboard/pull/795.

Unfortunately, Dashboard v0.3.0 is now bundled with Tekton Pipelines v0.8.0. So, we have the bug again.

I just verified that once the version of Tekton Pipelines is bumped up to v0.9.0 or greater, this bug will be resolved:
![Screen Shot 2020-01-16 at 11 50 12 AM](https://user-images.githubusercontent.com/16961496/72545875-be0c2a80-3857-11ea-93f4-e31f099b2c9d.png)
